src/host.cpp

```diff
@@ -274,6 +283,75 @@ namespace rhost {
             }
         }
 
+        void create_blob(const message& msg) {
+            assert(msg.name == "?CreateBlob");
+
+            if (msg.blob.size() != msg.blob_count) {
+                fatal_error("Incomplete blob creation message %s.", msg.id.c_str());
+            }
+
+            {
+                std::lock_guard<std::mutex> lock(received_blobs_mutex);
+                int blob_id = ++next_blob_id;
+                received_blobs[blob_id] = msg.blob;
+                respond_to_message(msg, static_cast<double>(blob_id));
+            }
+        }
+
+        int create_blob(const rhost::util::blob& blob) {
+            std::lock_guard<std::mutex> lock(received_blobs_mutex);
+            int blob_id = ++next_blob_id;
+            received_blobs[blob_id] = blob;
+            return blob_id;
+        }
+
+        const blob_slice get_blob_as_single_slice(int blob_id) {
+            std::map<int, rhost::util::blob>::iterator res = received_blobs.find(blob_id);
+            blob_slice data;
+
+            if (res == received_blobs.end()) {
+                return data;
+            }
+
+            for (blob_slice& slice : (*res).second) {
+                data.reserve(data.size() + slice.size());
```

Comment:
It seems like there could be a separate loop to compute the size first, and reserve the whole thing, before the loop that actually does the copying, to avoid any reallocs at all.